### PR TITLE
[TT-13184] Implement OAuth 2.0 Client Credentials Flow for GW authentication with upstream

### PR DESCRIPTION
**Pull request by andrei-tyk**
### **User description**
<details open>
  <summary><a href="https://tyktech.atlassian.net/browse/TT-13184" title="TT-13184" target="_blank">TT-13184</a></summary>
  <br />
  <table>
    <tr>
      <th>Summary</th>
      <td>Implement OAuth 2.0 Client Credentials for API Gateway Authentication with Upstream Server</td>
    </tr>
    <tr>
      <th>Type</th>
      <td>
        Story
        Story
      </td>
    </tr>
    <tr>
      <th>Status</th>
      <td>In Dev</td>
    </tr>
    <tr>
      <th>Points</th>
      <td>N/A</td>
    </tr>
    <tr>
      <th>Labels</th>
      <td>-</td>
    </tr>
  </table>
</details>

---

## Description

This PR introduces the OAuth 2.0 Client Credentials Flow into the Tyk API Gateway to enhance security and ensure authenticated communication between the Gateway and the Upstream server.

This PR introduces the OAuth 2.0 Client Credentials Flow within the Tyk API Gateway to enhance security for communication with the Upstream server. The Gateway will now authenticate itself by obtaining access tokens from the OAuth Authorization Server using the Client ID and Client Secret. Key changes include:

Token management: caching, refreshing, and handling expiration.
Secure storage of access tokens (hashed in Redis).
Improved error handling for token failures and upstream errors.
Configurable OAuth scopes and permissions in API definitions.
This ensures that only authorized requests are forwarded to the Upstream server.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Refactoring or add test (improvements in base code or adds test coverage to functionality)

## Checklist

- [ ] I ensured that the documentation is up to date
- [ ] I explained why this PR updates go.mod in detail with reasoning why it's required
- [ ] I would like a code coverage CI quality gate exception and have explained why

___

### **PR Type**
Enhancement, Tests

___

### **Description**
- Implemented OAuth 2.0 Client Credentials Flow for API Gateway authentication with upstream servers.
- Added `UpstreamOAuth` and `ClientCredentials` structs to manage OAuth2 configurations.
- Integrated OAuth2 support into existing `UpstreamAuth` structures and methods.
- Developed `UpstreamOAuth` middleware to handle OAuth2 authentication, including token caching and retrieval.
- Added tests for the new OAuth2 middleware to ensure functionality.
- Updated schema definitions to include OAuth2 properties.
- Refactored secret handling logic for better code organization.

___

### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><details><summary>10 files</summary><table>
<tr>
  <td>
    <details>
      <summary><strong>api_definitions.go</strong><dd><code>Add OAuth2 support to UpstreamAuth configuration</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

apidef/api_definitions.go

<li>Added OAuth2 configuration to <code>UpstreamAuth</code>.<br> <li> Introduced <code>UpstreamOAuth</code> and <code>ClientCredentials</code> structs.<br> <li> Updated <code>IsEnabled</code> method to include OAuth2.<br>

</details>

  </td>
  <td><a href="https://github.com/TykTechnologies/tyk/pull/6633/files#diff-9961ccc89a48d32db5b47ba3006315ef52f6e5007fb4b09f8c5d6d299c669d67">+43/-1</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>upstream.go</strong><dd><code>Integrate OAuth2 configuration in OAS UpstreamAuth</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

apidef/oas/upstream.go

<li>Added OAuth2 configuration to <code>UpstreamAuth</code>.<br> <li> Implemented <code>Fill</code> and <code>ExtractTo</code> methods for OAuth2.<br>

</details>

  </td>
  <td><a href="https://github.com/TykTechnologies/tyk/pull/6633/files#diff-7b0941c7f37fe5a2a23047e0822a65519ca11c371660f36555b59a60f000e3f4">+98/-3</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>schema.go</strong><dd><code>Extend schema to include OAuth2 properties</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

apidef/schema.go

- Added OAuth2 properties to the schema.

</details>

  </td>
  <td><a href="https://github.com/TykTechnologies/tyk/pull/6633/files#diff-f8a37bb370eb6fe20063786a5e6ea3d85a5c91d8e289f0b3e045830c4d322095">+34/-0</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>api_loader.go</strong><dd><code>Append UpstreamOAuth middleware to API loader</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

gateway/api_loader.go

- Appended `UpstreamOAuth` middleware to the chain.

</details>

  </td>
  <td><a href="https://github.com/TykTechnologies/tyk/pull/6633/files#diff-cdf0b7f176c9d18e1a314b78ddefc2cb3a94b3de66f1f360174692c915734c68">+1/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>event_system.go</strong><dd><code>Add EventUpstreamOAuthMeta for OAuth events</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

gateway/event_system.go

- Added `EventUpstreamOAuthMeta` for OAuth events.

</details>

  </td>
  <td><a href="https://github.com/TykTechnologies/tyk/pull/6633/files#diff-d56e22d4f1b8d2e91bb643d30e678a3819691a18bfae8506b10e0af8dc279a0e">+6/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>middleware.go</strong><dd><code>Implement emitUpstreamOAuthEvent method</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

gateway/middleware.go

- Added `emitUpstreamOAuthEvent` method for OAuth events.

</details>

  </td>
  <td><a href="https://github.com/TykTechnologies/tyk/pull/6633/files#diff-703054910891a4db633eca0f42ed779d6b4fa75cd9b3aa4c503e681364201c1b">+15/-0</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>mw_oauth2_auth.go</strong><dd><code>Implement UpstreamOAuth middleware for OAuth2 authentication</code></dd></summary>
<hr>

gateway/mw_oauth2_auth.go

<li>Implemented <code>UpstreamOAuth</code> middleware for OAuth2 authentication.<br> <li> Added token caching and retrieval logic.<br>

</details>

  </td>
  <td><a href="https://github.com/TykTechnologies/tyk/pull/6633/files#diff-a90347c3ad28f06a7bd1c5554ce63448774cb486cf4e9961af2323423ce8209d">+244/-0</a>&nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>rpc_backup_handlers.go</strong><dd><code>Refactor secret padding logic</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

gateway/rpc_backup_handlers.go

- Refactored secret padding logic into `getPaddedSecret`.

</details>

  </td>
  <td><a href="https://github.com/TykTechnologies/tyk/pull/6633/files#diff-69d9cb8df2bd4296a8e5e5d769009a09bd61ca65b7dbcbf29751af92698bd9ce">+5/-2</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>server.go</strong><dd><code>Add UpstreamOAuthCache to Gateway struct</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

gateway/server.go

- Added `UpstreamOAuthCache` to `Gateway` struct.

</details>

  </td>
  <td><a href="https://github.com/TykTechnologies/tyk/pull/6633/files#diff-4652d1bf175a0be8f5e61ef7177c9666f23e077d8626b73ac9d13358fa8b525b">+2/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>x-tyk-api-gateway.json</strong><dd><code>Extend X-Tyk-UpstreamAuthentication with OAuth2 configuration</code></dd></summary>
<hr>

apidef/oas/schema/x-tyk-api-gateway.json

- Added OAuth2 configuration to `X-Tyk-UpstreamAuthentication`.

</details>

  </td>
  <td><a href="https://github.com/TykTechnologies/tyk/pull/6633/files#diff-78828969c0c04cc1a776dfc93a8bad3c499a8c83e6169f83e96d090bed3e7dd0">+38/-1</a>&nbsp; &nbsp; </td>

</tr>                    
</table></details></td></tr><tr><td><strong>Tests</strong></td><td><details><summary>1 files</summary><table>
<tr>
  <td>
    <details>
      <summary><strong>mw_oauth2_auth_test.go</strong><dd><code>Add tests for UpstreamOAuth middleware</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

gateway/mw_oauth2_auth_test.go

- Added tests for `UpstreamOAuth` middleware.

</details>

  </td>
  <td><a href="https://github.com/TykTechnologies/tyk/pull/6633/files#diff-4bbf88199c7bb23cea3a299d25c6074d7c61c481a48e51b3fa7d5f999ab581a1">+127/-0</a>&nbsp; </td>

</tr>                    
</table></details></td></tr></tr></tbody></table>

___

> 💡 **PR-Agent usage**: Comment `/help "your question"` on any pull request to receive relevant information

### CODE REVIEW

File: apidef/api_definitions.go [hunk 1]
```diff
@@ -793,6 +802,39 @@ type UpstreamBasicAuth struct {
 	HeaderName string `bson:"header_name" json:"header_name"`
 }
 
+type UpstreamOAuth struct {
```
Comment (titpetric) [inline apidef/api_definitions.go:806]: Missing godoc on type, same below

File: apidef/oas/upstream.go [hunk 2]
```diff
@@ -556,7 +557,10 @@ type UpstreamAuth struct {
 	// Enabled enables upstream API authentication.
 	Enabled bool `bson:"enabled" json:"enabled"`
 	// BasicAuth holds the basic authentication configuration for upstream API authentication.
-	BasicAuth *UpstreamBasicAuth `bson:"basicAuth,omitempty" json:"basicAuth,omitempty"`
+	BasicAuth *UpstreamBasicAuth `bson:"basicAuth" json:"basicAuth"`
```
Comment (titpetric) [inline apidef/oas/upstream.go:560]: Why does this remove omitempty?

File: gateway/mw_oauth2_auth_test.go [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+
+	"github.com/TykTechnologies/tyk/apidef"
+	"github.com/TykTechnologies/tyk/header"
+	"github.com/TykTechnologies/tyk/test"
+)
+
+func TestUpstreamOauth2(t *testing.T) {
+
+	tst := StartTest(nil)
+	t.Cleanup(func() {
```
Comment (titpetric) [inline gateway/mw_oauth2_auth_test.go:21]: t.Cleanup(tst.Close)?

**Review by titpetric — APPROVED** [review]

File: apidef/oas/upstream.go (hunk 2 above)
Comment (andrei-tyk) [reply apidef/oas/upstream.go]: Mistake, added back

File: gateway/mw_oauth2_auth_test.go (hunk 3 above)
Comment (andrei-tyk) [reply gateway/mw_oauth2_auth_test.go]: Fixed

File: apidef/api_definitions.go (hunk 1 above)
Comment (andrei-tyk) [reply apidef/api_definitions.go:807]: Fixed

**Pull request merged**